### DISCUSSION TITLE
How to pre-define symbols?

### DISCUSSION THREAD

**Discussion by frozax** [microsoft/pylance-release#1149] [Q&A] [answered]
I'm using vscode and python for https://py.processing.org/. 
In this environnement, some symbols are available without being imported. For example, `frameRate` and `circle` are valid methods, and `frameCount` and `width` are variables.
How could I give a list of those symbols (eventually with a prototype for function) to Pylance to avoid `reportUndefinedVariable`?

Thanks for your help!

**Comment by erictraut** [comment, marked as answer]
While Python allows code to "insert" symbols into the namespace of another module, this practice is very fragile, and I would highly discourage library authors from using it. It potentially overwrites symbols of the same name in other modules, and it makes code difficult to follow (both for static code tools and for humans). Symbols should be imported explicitly to avoid namespace collisions. 

If you are the consumer of a library that implements this dangerous and ill-advised behavior, you have a few options:

1. Although the symbol is already implicitly "inserted" into your module's namespace, use a proper _explicit_ import statement to import the symbol again. Pylance and other tools (and readers of your code) will then be aware of where that symbol originates and how it came to be part of your module's namespace.

```python
from processing import frameRate, frameCount
```

2. Conditionally define the types of these symbols within a `TYPE_CHECKING` block within your module:
```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    def frameRate(a: int, b: str) -> int: ...
    frameCount: int
```

The first of these two options is preferable if it works.

**Reply by frozax** [reply]
Thanks Eric for the reply.

There is actually no `processing` stuff to import. Everything is hidden with a specific interpreter (yes, it's a bit hacky: https://py.processing.org/tutorials/python-jython-java/). If I create a fake module with a `frameCount` and import it in my code for example, it override the actual one.

I basically would like to tell pylance: "here is a list of valid symbols, believe me" ;)

Also, I'm not sure to understand your 2nd solution?

**Reply by erictraut** [reply]
Oh, that's really hacky. Yikes!!!

You can create a fake module and import it under the `TYPE_CHECKING` hint. That way, it won't execute at runtime but static type checkers like Pylance will make use of it during static analysis.

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    # Provide types for symbols that are auto-imported by "processing"
    from _processing import frameRate, frameCount
```

You can define the symbols in a type stub file located in the root of your project.
```python
# _processing.pyi

def frameRate(a: int, b: str) -> int: ...
frameCount: int
# etc.
```
